### PR TITLE
Multithreading leaf evaluations

### PR DESCRIPTION
**Pull request by phiedw**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*
No

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
Feature

**What is the current behavior?** *(You can also link to an open issue here)*
Leaves are run successively

**What is the new behavior (if this is a feature change)?**
Leaves are run simultaneously

**Does this PR introduce a breaking change?** *(What changes might users need to make in their application due to this PR?)*
No

**Other information**:

(if any of the questions/checkboxes don't apply, please delete them entirely)

### CODE REVIEW

File: ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/process/search_tree/Tree.java [hunk 1]
```diff
@@ -30,6 +32,8 @@
  * routing among the leaves in order to converge as quickly as possible to a local
  * minimum of the objective function.
  *
+ * The leaves of a same depth cen be evaluated simultaneously.
```
Comment (terriervik) [inline ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/process/search_tree/Tree.java:35]: Typo in "can"

File: ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/process/search_tree/Tree.java [hunk 2]
```diff
@@ -63,20 +67,21 @@ private Tree() {
...
+            List<Leaf> successfulLeaves = generatedLeaves.stream().filter(leaf -> leaf.getStatus() == Leaf.Status.EVALUATION_SUCCESS).collect(Collectors.toList());
 
             hasImproved = false;
-            for (Leaf currentLeaf: generatedLeaves) {
+            for (Leaf currentLeaf: successfulLeaves) {
                 if (improvedEnough(optimalLeaf.getCost(crac), currentLeaf.getCost(crac), searchTreeRaoParameters)) {
                     hasImproved = true;
+                    optimalLeaf.deletePostOptimResultVariant(crac);
                     depth = depth + 1;
                     optimalLeaf.deletePostOptimResultVariant(crac);
```
Comment (terriervik) [inline ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/process/search_tree/Tree.java:86]: You do this twice here, it doesn't seem necessary.
optimalLeaf.deletePostOptimResultVariant(crac);

File: ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/process/search_tree/Tree.java (hunk 1 above)
Comment (phiedw) [reply ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/process/search_tree/Tree.java]: fixed

File: ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/process/search_tree/Tree.java (hunk 2 above)
Comment (phiedw) [reply ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/process/search_tree/Tree.java]: indeed

File: ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/process/search_tree/Tree.java [hunk 3]
```diff
@@ -63,18 +67,18 @@ private Tree() {
 
         while (doNewIteration(searchTreeRaoParameters, hasImproved, optimalLeaf.getCost(crac), depth)) {
             Set<NetworkAction> availableNetworkActions = crac.getNetworkActions(network, crac.getPreventiveState(), UsageMethod.AVAILABLE);
-            List<Leaf> generatedLeaves = optimalLeaf.bloom(availableNetworkActions);
+            final List<Leaf> generatedLeaves = optimalLeaf.bloom(availableNetworkActions);
```
[the remark this answers is not in the document]
Comment (phiedw) [reply ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/process/search_tree/Tree.java:70]: it's used in a lambda function

**Pull request merged**